### DISCUSSION TITLE
Factory function used in the class Graph.

### DISCUSSION THREAD

**Discussion by zjumhy97** [networkx/networkx#6123] [Q&A] [answered]
In the source code of the class Graph, there is a 
class Graph():
     node_dict_factory = dict

     def __init__(self):
          self.node_dict_factory = self.node_dict_factory
          ......
          self._node = self.node_dict_factory

Why use this approach, is this related to design pattern? So much thanks!

**Comment by rossbar** [comment, marked as answer]
Yes, the factory function design approach is intentional. Organizing things in this way allows a user who wants to subclass one of the builtin networkx graph classes to inject a custom dict-like object in the structure of nested dicts. There is a concrete example demonstrating this utility to create a `ThinGraph` subclass, which can save a lot of memory at the expense of disallowing per-edge attributes. Check out the ["Subclassing (Advanced)" section of the Graph docstring](https://github.com/networkx/networkx/blob/73afc6768021f46279d05e6b8a8e722bffac1d79/networkx/classes/graph.py#L226-L303).

For a more general overview of the factory function design pattern, I'd recommend the Wikipedia article or perhaps checking out the source code for other projects that make heavy use of factory functions such as `pytest` or `flask`.

**Reply by zjumhy97** [reply]
Thank u sooooo much! It really helps me! I will go to learn how this utility works!
